Feature/afloat pallet

Dreyhh opened the pull request:
# Transactions

## Overview

This PR introduces a first approach to sell transactions.

Each Offer now contains an associated array of Transactions.
Each Transaction has only one offer_id associated with it.

Let's suppose that Bobs wants to take a Sell Order (partially or fully) from Alice.

First, Bob must call:
```rust
start_take_sell_order(offer_id, tax_credit_amount)
```
This will create a Transaction, which means that Alice's Sell Offer now has a Transaction associated with it.
(at this point the Transaction is confirmed by the Buyer but not so by the Seller).

If Alice is interested in this Transaction, she must confirm it:
```rust
confirm_sell_transaction(transaction_id)
```
Now the Transaction is confirmed by both the Seller and the Buyer. but it's still uncompleted, to complete the Transaction, the Buyer must call:
```rust
 finish_take_sell_transaction(transaction_id)
```
Upon a Transaction's completion, funds and tax credits will be sent to the Seller and Buyer respectively.

## Test coverage

Although unit tests have been written to cover basic functionality more exhaustive ones are needed.

Dreyhh committed:
Fixes marketplaces creation, the mapped assed is now created separately

Dreyhh committed:
Introduces sell transactions, improves error handling, updates unit tests

Dreyhh committed:
Change state of transaction when completed

tlacloc commented on pallets/afloat/src/functions.rs:
```diff
@@ -252,23 +255,25 @@ impl<T: Config> Pallet<T> {
+		let collection_id = AfloatCollectionId::<T>::get().ok_or(Error::<T>::CollectionIdNotFound)?;
+		let transactions = TransactionBoundedVec::default();
 
 		let offer_id = pallet_gated_marketplace::Pallet::<T>::do_enlist_sell_offer(
```
GM orders are not linked 1 to 1 with Afloat orders, so please remove this line

tlacloc commented on pallets/afloat/src/functions.rs:
```diff
@@ -295,23 +301,25 @@ impl<T: Config> Pallet<T> {
+		let collection_id = AfloatCollectionId::<T>::get().ok_or(Error::<T>::CollectionIdNotFound)?;
+		let transactions = TransactionBoundedVec::default();
 
 		let offer_id = pallet_gated_marketplace::Pallet::<T>::do_enlist_buy_offer(
```
same with this one

tlacloc commented on pallets/afloat/src/lib.rs:
```diff
@@ -159,6 +200,16 @@ pub mod pallet {
+			let min_balance: T::Balance = 1u32.into();
+
+			// Create asset
+			pallet_mapped_assets::Pallet::<T>::create(
```
add a boolean in initial setup related the creation of the MA asset, something as 
```rust
pub fn initial_setup(
			origin: OriginFor<T>,
			creator: T::AccountId,
			admin: T::AccountId,
            create_asset: bool,
		) 
```
Just to make sure we can reuse an asset even if we delete all the storage maps

tlacloc commented on pallets/afloat/src/lib.rs:
```diff
@@ -193,7 +244,8 @@ pub mod pallet {
 			let who = ensure_signed(origin.clone())?;
-			ensure!(Self::is_admin_or_owner(who), Error::<T>::Unauthorized);
+			let is_admin_or_owner = Self::is_admin_or_owner(who.clone())?;
+			ensure!(is_admin_or_owner, Error::<T>::Unauthorized);
```
I like more the idea of just a Sudo account can kill the storage

tlacloc commented on pallets/afloat/src/lib.rs:
```diff
@@ -159,6 +200,16 @@ pub mod pallet {
 	{
 			ensure_signed(origin.clone())?;
```
the origin needs to be a sudo account

tlacloc commented on pallets/afloat/src/lib.rs:
```diff
@@ -255,46 +307,45 @@ pub mod pallet {
 		#[pallet::call_index(5)]
 		#[pallet::weight(Weight::from_ref_time(10_000) + T::DbWeight::get().reads_writes(1,1))]
-		pub fn accept_offer(
+		pub fn start_take_sell_order(
```
It's okay for now, but I prefer the idea of having an extrinsic that receives an order, and then the backend/pallet determines the flow accordingly. For example, we can process it as a sell order if the type is 'sell,' and so on.

tlacloc commented on pallets/afloat/src/lib.rs:
```diff
@@ -255,46 +307,45 @@ pub mod pallet {
 		#[pallet::call_index(6)]
 		#[pallet::weight(Weight::from_ref_time(10_000) + T::DbWeight::get().reads_writes(1,1))]
-		pub fn take_sell_order(
+		pub fn confirm_sell_transaction(
```
same as this one

tlacloc reviewed (changes requested):
In general pretty good, thank you for developing the unit tests for this changes

Dreyhh committed:
Removes GM offers creation, and adds sudo origin checks for initial setup and kill storage

Dreyhh committed:
Adds afloat-specific functions to mint and burn mapped-assets and extrinsic to add new afloat administrators

tlacloc committed:
✨ feat(types.rs): add new OfferStatus enum variants to improve semantics and readability of the code. Specifically, CREATED, TF_SUBMITTED, TF_APPROVED, APPROVED, and FILLED were added. The default value was also changed from TF_PENDING_SIGNATURE to CREATED.

tlacloc committed:
Fix formating issues

tlacloc committed:
Merge branch 'develop' into feature/afloat-pallet

tlacloc committed:
🔨 refactor(functions.rs): add cid_creator field to User struct to track the creator of the user's CID
🔨 refactor(functions.rs): change SignUpArgs and UpdateUserArgs to include cid_creator field
🔨 refactor(functions.rs): change do_edit_user and do_admin_edit_user functions to include cid_creator field

tlacloc committed:
🚀 feat(lib.rs): add SessionKeys struct with aura and grandpa keys to the opaque module to enable session key management and validation. Also, expose Block, Header and BlockId types to allow for agnostic syncing of the network through upgrades.
